[Add] Authentication & Validation

DalioKim opened the pull request:
# 주요 변경사항
	- Authentication View & ViewModel 추가
	- 인증시 입력한 정보에 대한 정규표현식 검사
	- 정규표현식 검사 결과를 표시하는 FieldStyle 추가

# 중점적으로 봐줬으면 하는 부분
	- UI & 비즈니스 로직 분리도 
	- AuthenticationView 프로퍼티 및 메서드 분리 및 선언 구조
	- AuthenticationView 네임스페이스
	- ViewModel Bind & didValidate 로직
	- Authentication Ux Flow
	- 네이밍

# 작업 예정사항
	- AuthenticationView TDD
	- Authentication Service
	- Authentication flow as AppManager

DalioKim committed:
FEAT: AuthenticationView
1. 로그인 및 회원가입을 위한 입력필드등 UI 구현
2. 리터럴값 NameSpace에 Constant로 지정

DalioKim committed:
FEAT: AuthenticationViewModel 추가

1.String+isValid - 이메일, 비밀번호 정규식 추가
2.AuthenticationView Field - alert 연산자를 통해 잘못된 입력에 대한 경고창 띄우기 추가
3.AuthenticationViewModel - 바인딩 변수, 유효성 검사할 변수 구현
4.bindValidation 함수를 통해 바인딩 변수를 구독하고 유효성 상태 변환

DalioKim committed:
REFACTOR:
1 String+isValid 파일: 계산속성 어순 변경
2 AuthenticationViewModel 파일
2-1) isNotEmailValid, isNotPasswordValid 을 isEmailValid, isPasswordValid 로 변경
2-2) bindValidation 메서드에서 throttle 로직 제거

FEAT:
1. ValidationFieldModifier 구현 - 유효성 검사 연산자

DalioKim committed:
REFACTOR:

1) ValidationFieldModifier:
	1-1) 메서드 네이밍 수정
	1-2) 생성자에 borderColor 추가

2) AuthenticationView
	2-1) 프로퍼티, 메서드 선언 위치 분리 및 변경
	2-2) 네임스페이스 변경

FEAT:

1) AuthenticationViewModel:
	1-1) isInputsValid 프로퍼티 추가
	1-2) didValidate 메서드 추가

2) AuthenticationView:
	1-1) isInputsValid sink 로직 추가

hanjaesueng commented on GatheRunner/Presentaion/Authentication/AuthenticationView.swift:
```diff
@@ -0,0 +1,167 @@
+        }
+    }
+
+    private var fieldPicker: some View {
+        Picker(Content.Label.empty, selection: $isSignIn) {
+            Text(Content.Label.signIn).tag(true)
+            Text(Content.Label.signUp).tag(false)
+        }
+        .pickerStyle(SegmentedPickerStyle())
+        .padding()
+    }
+
+    private var emailField: some View {
+        TextField(Content.Label.email, text: $viewModel.email)
+            .keyboardType(.emailAddress)
+            .disableAutocorrection(true)
+            .frame(width: Size.width, height: Size.height, alignment: .center)
+            .asValidationFieldStyle(isValid: $viewModel.isEmailValid)
+    }
+
+    private var passwordField: some View {
+        SecureField(Content.Label.password, text: $viewModel.password)
+            .frame(width: Size.width, height: Size.height, alignment: .center)
+            .asValidationFieldStyle(isValid: $viewModel.isPasswordValid)
+    }
+}
```
질문
--------------------------------------------
여기서 focus를 적용하실 예정이신가요??

DalioKim replied:
포커스는 asValidationFieldStyle을 통해 적용되어 있습니다.

hanjaesueng commented on GatheRunner/Presentaion/Authentication/AuthenticationView.swift:
```diff
@@ -0,0 +1,167 @@
+
+// MARK: - AuthenticationView
+
+struct AuthenticationView {
```
의견
--------------------------------------------
AuthenticationView 를 보면 한번에 파악하기가 좀 힘든 경향이 있습니다. 잘 정리해 주셨지만 일단 양이 많았고 정리하고 구분지어 주신것을 따라가다보니 코드 의도파악과 분석까지 조금 시간이 걸렸습니다. Dalio만의 규칙으로 정리되어 있는걸까요??

hanjaesueng replied:
"Dalio만의 규칙으로 정리되어 있는걸까요??" 질문이 약간 예민할 수 있을거 같아서 다시 코멘트 달자면 세세히 나누신게 혹시 참고한 레퍼런스나 그런게 있는지 질문드립니다.

DalioKim replied:
extension을 활용해서 private 타입, 메서드, 네임스페이스 등등 성격별로 분리하여 사용하는 경우가 있어, 적용해봤습니다.

참고 레퍼런스
[링크 1](https://blog.flaviocaetano.com/post/a-better-way-to-organize-swift-classes/)
[링크 2](https://medium.com/@nimjea/extensions-in-swift-d87d6ed0c6fd)

또한 피라미드구조 혹은 body 부분의 코드가 길어지는 것을 피하기 위해 UI 로직을 분리하고 있습니다.
[링크](https://medium.com/@_DandyLyons/7-ways-to-organize-swiftui-code-e786307d3916)

분리의 순서와 기준은 좀 더 보편화된 기준에 따라 배치하거나 불필요한 분리는 없애도록 하겠습니다.

DalioKim replied:
그리고 배치 순서는 제가 지금 자료를 찾지 못하겠는데, 이번에 작성한것과 같이 private 변수와 body를 구분해서 사용하는 자료를 봐서 그 형식으로 배치하였고, 

나머지 서브뷰의 UI로직과 네임스페이스와 메서드는 상단에 배치하면 불필요한 정보를 처음부터 많이 제공한다고 생각하여 하단에 배치하였습니다.

DalioKim committed:
REFACTOR:

ValidationFieldModifier
1.ValidationFieldModifier cornerRadius, lineWidth 네임스페이스로 기본 값 지정
2. asValidationFieldStyle overloading하여 기본값 유지하는 생성자와 변경하는 생성자 구분
3. 리터럴 값 네임스페이스변수로 변경

AuthenticationView
1. extension 그룹화 기준 및 배치 변경

DalioKim committed:
REFACTOR:

ValidationFieldModifier
1. namespace fileprivate으로 접근 제한자 수정 후 asValidationFieldStyle의 파라미터상에서 값 설정
2. overloadiing한 asValidationFieldStyle 메서드 삭제 및 typealias DefaultSize 추가
3. didValidate() 삭제 후, onSubmit에서 UI 로직 구현

AuthenticationViewModel
1. didValidate 메서드 isInputsValid으로 네이밍 변경

FIX: AuthenticationView의 isValiid 프로퍼티 네이밍 isValid로 변경하여 오타 수정

DalioKim commented:
# 주요 변경사항
	- AuthenticationView: extension 그룹화 기준 및 배치 변경
	- AuthenticationViewModel: didValidate 메서드네이밍 validatedInputs 으로 변경
	- ValidationFieldModifier: onSubmit UI 로직 변경 및 리터럴 값 네임스페이스로 변경

# 중점적으로 봐줬으면 하는 부분
	- 네임스페이스 적합여부
	- 변경한 UI 로직 및 네이밍
	- AuthenticationView 변경한 배치 및 그룹화 기준

wh5929 commented on GatheRunner/Helper/UI/Style/FieldStyle.swift:
```diff
@@ -0,0 +1,56 @@
+    @ViewBuilder
+    func asValidationFieldStyle(
+        isValid: Binding<Bool>,
+        witBorderColor borderColor: Color = .gray,
```
wit 오타같네요.
